### PR TITLE
feat(rstest): add no-conditional-tests rule

### PR DESCRIPTION
**Pull request by elecmonkey**
## Summary

Report a test or suite registration that is written inside an `if` statement's `then` or `else` branch, since a registration that only runs on some executions is a test whose coverage silently shrinks instead of failing. No fix is offered.

The call resolves through the shared `ParseFnCall`, so only a real Rstest `test` or `describe` registration is reported — hooks, a same-named export from another test framework, and a local variable that merely shares the name are all left alone. The walk from the registration call up to its enclosing scope stops at the first deferred-execution boundary — a function, arrow, method, constructor, or accessor body, a parameter default, or an instance field initializer — or at the source file. It never follows a function to its call sites, so a registration wrapped in a helper is not reported at all, and a nested pair such as `if (x) { describe('a', () => { test('b', fn) }) }` reports only the outer `describe`. Positions that run while a class is defined — static blocks, static field initializers, computed member names, decorators — are not boundaries, so a registration in one stays conditional on an enclosing `if`. A call sitting in the `if`'s own condition, such as `if (test('a')) {}`, is not reported — it runs unconditionally. Registered in the rstest plugin but left out of the `recommended` preset, matching upstream.

## Credits

Port from [`eslint-plugin-vitest`](https://github.com/vitest-dev/eslint-plugin-vitest)'s `no-conditional-tests`. `eslint-plugin-jest` has no counterpart, so there is no second baseline to cross-check against.

## Intentional differences from upstream

- **The registration is resolved instead of name-matched.** Upstream reports any `Identifier` literally spelled `test`, `it`, or `describe`, so an ordinary variable with one of those names is reported even though it has nothing to do with the test framework. This port only reports a call it can resolve to a Rstest registration through `ParseFnCall`, which also lets it follow imports from `@rstest/core` and `@rstest/playwright`, `require` forms, `import.meta.rstest` members, globals, and local aliases, while staying silent on a shadowed name or a same-named API from Vitest, Jest, or `node:test`.

- **The parent walk follows the enclosing scope instead of a fixed depth.** Upstream checks whether the identifier's parent four levels up is exactly an `IfStatement`, which assumes the registration is a direct expression statement inside a braced block. `if (cond) test('a', fn)`, written without braces, puts the call three levels below the `if` instead of four, and upstream misses it. This port walks up to the first function boundary or the source file instead, so the shape and spacing of the intervening statements no longer matter, and the same walk naturally reports only the outermost registration of a nested pair.

## Related Links

Related #935

## Checklist

- [x] Tests updated (or not required).
- [x] Documentation updated (or not required).

### CODE REVIEW

File: internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+// `if (test('a')) {}` runs the call unconditionally — so the walk continues
+// past it rather than reporting, and an outer `if` can still catch it.
+func registeredUnderIfBranch(node *ast.Node) bool {
+	child := node
+	for {
+		parent := child.Parent
+		if parent == nil || parent.Kind == ast.KindSourceFile {
+			return false
+		}
+		if ast.IsFunctionLikeOrClassStaticBlockDeclaration(parent) {
```
Comment (swwind) [inline internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.go]: `ast.IsFunctionLikeOrClassStaticBlockDeclaration` conflates distinct execution phases into a single boundary. Static blocks, computed property names, and method decorators within a class are executed when the class definition is evaluated; thus, the registration within `if (flag) { class C { static { test(...) } [test(...)]() {} @dec(test(...)) m() {} } }` is indeed controlled by the `if` statement, yet the current logic returns early and fails to report it. Conversely, instance fields like `p = test(...)` execute only upon instance construction; the current traversal passes through the field declaration and incorrectly attributes it to the `if` statement enclosing the class declaration. Please redefine the boundaries based on the type of class member and the position of the current child node: traversal should halt for function/method bodies, parameter default values, and instance field initializers, but should continue for static blocks, static initializers, computed property names, and decorators; additionally, please add corresponding regression tests.

Comment (elecmonkey) [reply internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.go]: Confirmed — thanks for the precise breakdown. Fixed in 5346b3b3b5: the boundary check now distinguishes deferred-execution positions (function/method/constructor/accessor bodies, parameter default values, instance field initializers) from synchronous class-evaluation-time positions (static blocks, static field initializers, computed member names, decorators). Added regression tests for all four non-boundary shapes plus the instance-field-initializer boundary case. Also corrected the shared batch spec's boundary definition, since it previously listed class static block as a stopping point.

**Review by swwind — COMMENTED** [review]
isDeferredExecutionBoundary treats every instance PropertyDeclaration and every method Parameter as deferred. That misses registration expressions which execute during class evaluation rather than during instance construction or a later method call:

```js
if (x) {
   class C {
    [test("a", () => {})] = 1;
  }
}

if (x) {
  class C {
    @dec(test("a", () => {}))
    p = 1;
  }
}

if (x) {
  class C {
    method(@dec(test("a", () => {})) value: unknown) {}
  }
}
```

All three registrations are conditional but currently go unreported. Please distinguish a field initializer / parameter default (deferred) from a computed name or decorator (synchronous class evaluation).

----

The shared registration parser does not recognize a comma-wrapped test callee:

```js
if (x) {
  (0, test)("a", () => {});
}
```

This still invokes the real Rstest test registration conditionally, but the rule emits no diagnostic. Consider unwrapping a comma expression whose right operand is the resolved registration reference, or document this intentional parser limitation.

**Comment by elecmonkey** [comment]
@swwind Fixed both points in 99d9169781.

- Made class-member execution boundaries position-sensitive, so computed field names, field decorators, and parameter decorators are treated as class-evaluation-time expressions, while instance initializers and parameter defaults remain deferred.
- Updated the shared Rstest call parser to unwrap comma expressions through their right operand, including imports, aliases, hooks, and `import.meta.rstest` calls.
- Added dedicated class-timing regressions and updated the shared parser and `no-hooks` contract tests.

The affected Go tests, binary build, JS integration test, typecheck, lint, spell check, formatting check, and targeted Go lint all pass.

File: internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		Id:          "noConditionalTests",
+		Description: "Avoid using if conditions in a test",
+	}
+}
+
+var NoConditionalTestsRule = rule.Rule{
+	Name:   "rstest/no-conditional-tests",
+	Schema: rule.EmptyArraySchema,
+	Run: func(ctx rule.RuleContext, options []any) rule.RuleListeners {
+		analysis := rstestUtils.GetRstestCallAnalysis(ctx)
```
Comment (swwind) [inline internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.go:21]: [P2] Preserve registration resolution in source-only programs

This rule does not require type information, but `GetRstestCallAnalysis` loses binding resolution when `ctx.TypeChecker` is nil. In a focused source-only `.ts`/`.js` multi-file run, `import { test } from 'vitest'; if (flag) { test(...) }` and a locally declared `test` were incorrectly reported as Rstest globals, while renamed and namespace imports from `@rstest/core` were missed. Please use the file-local reference store as the fallback when resolving the root symbol, or require type information for this rule; otherwise the source discrimination promised by this port reverses in both directions outside checker-backed programs.

File: internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+The rule does not carry an enclosing `if` through code that runs later. A
+registration in a function or method body, a parameter default, or an instance
+field initializer is attributed to the later call or construction rather than
+to an `if` around its declaration. Class static blocks, static field
+initializers, computed member names, and decorators run while the class is
+defined, so registrations in those positions remain conditional on an
+enclosing `if`.
+
+A registration wrapped in a helper is therefore attributed to wherever that
```
Comment (swwind) [inline internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.md]: [P3] Do not claim helper call-site attribution

The implementation stops when it crosses a function body and does not inspect where that function is called. Consequently, `function register() { test("a", fn) } if (flag) { register() }` emits no diagnostic, as does an equivalent IIFE inside the branch. Please reword this paragraph to say only that conditions do not cross deferred-execution boundaries, or implement the claimed call-site attribution.

File: internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.go (hunk 2 above)
Comment (elecmonkey) [reply internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.go:21]: Fixed in a62280f0.

`resolveRstestRootSymbol` now falls back to `ctx.Refs.ResolveInFile` when no TypeChecker is available, and `GetRstestCallAnalysis` carries `Refs` into the cached analysis context. The binder scope walk places imports and declarations authored in the file, which covers every discrimination the parser promises; bindings declared outside the file were never registrations.

The degradation lived in the shared call parser, so every rstest rule resolving through `ParseFnCall` picks up the fix, not only this one.

Added `TestNoConditionalTestsResolvesInSourceOnlyProgram`, which lints a checker-free `NewFromRoots` Program and asserts that `import { test } from 'vitest'` and a local `const localTest` are not reported, while a renamed import, a namespace import, and a plain import from `@rstest/core` are.

File: internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.md (hunk 3 above)
Comment (elecmonkey) [reply internal/plugins/rstest/rules/no_conditional_tests/no_conditional_tests.md]: Reworded in a62280f0. The doc no longer claims call-site attribution: it now says the rule does not follow a function to its call sites, so a registration in a helper is never reported, whether or not the helper is only ever called from inside an `if`. The paragraph above it likewise says such registrations are left alone, rather than being "attributed to the later call or construction".

**Review by swwind — APPROVED** [review]
